moontool.1
.TH MOONTOOL 1 "15 JULY 2024"
.UC 4
.SH NAME
moontool \- Werewolf Early Warning System
.SH SYNOPSIS
.B moontool
[-t UNIXTIME] [-f FORMATTING] [-h]
.SH DESCRIPTION
.B Moontool
is a SunView application which displays information about the moon in
real time.  When closed (iconic) it displays a graphical representation
of what the moon would look like right now if you were to go outside and
look at it. (Go on, try it, the fresh air may do you some good).  When
the window is open, the same image of the moon is displayed in the upper
right of the window, along with more verbose information about both the
moon and the sun (the bright yellow thing in the sky, not the thing you're
staring at right now.  It's not a good idea to stare at the bright yellow
thing).  The open window display is updated approximately once
per second, the icon is updated every two minutes.
.PP
.B Moontool
will run on both monochrome and color Suns.  Unless explicity told to
run in either the monochrome
.I (-m)
or the color
.I (-c)
mode,
.B moontool
will automatically select the appropriate mode.  On color displays,
the moon's face is rendered in eight colors in both the icon and the
open window.  The darkened portion of the moon is visible on color displays
as dark shades of blue.  On monochrome displays, only the illuminated
portion of the face is drawn in.
.PP
If
.B moontool
is run with the
.I -b
option, a one pixel border will be drawn around the moon's image
when the window is iconic.  This may improve the aesthetic appeal
of the icon, depending on your choice of desktop background and/or
the arrangement of windows and icons on the screen.  The border is
white on monochrome displays, light red on color.
.PP
As an extra added bonus for users of
color displays, the Apollo 11 Commemorative Red Dot has been added to the
face of the moon at the approximate location of Tranquility Base, to
mark the 20th anniversary of An Historic Event.
.PP
The image of the moon's face used by
.B moontool,
especially the color version, is quite accurate.  It was rendered on an
Amiga by Joe Hitchens, an artist of no small talent, from an illustration
in an astronomy textbook.
.PP
If invoked with the
.I -t
option,
.B moontool
will start up in test mode.  In test mode, the tool warps through time,
updating the display as fast as possible, skipping approximately one
hour per cycle.  A menu is available on the tool, for both the icon and
the open window, which will let you switch in and out of test mode at will.
This is what's known as a cheap thrill, and can provide endless hours of
entertainment (if you're easily entertained).
.PP
For those who are lycanthropically inclined, the open window tells
you when the next full moon will occur.  This may or may not be something
you'll want to know.
.SH TEDIOUS TECHNICAL CREDITS
Believe it or not,
.B moontool
uses some actual science and even authentic mathematics.  If you're into
that sort of thing, you might like to know that the following books
were used by the original author to create
.B moontool:
.PP
"Practical Astronomy With Your Calculator" by Peter Duffett-Smith,
Second Edition, Cambridge University Press, 1981.
.PP
"Astronomical Formulae for Calculators" by Jean Meeus, Third Edition,
Willmann-Bell, 1985.
.PP
"Planetary Programs and Tables from -4000 to +2800" by Pierre Bretagnon
and Jean-Louis Simon, Willmann-Bell, 1986.
.PP
"Celestial BASIC" by Eric Burgess, Revised Edition, Sybex, 1985.
.PP
This one was used by Joe to render the moon image and by me to locate
the Apollo 11 landing site:
.PP
"Astronomy: From the Earth to the Universe" by Jay M. Pasachoff,
Second Edition, Saunders College Publishing, 1983
.SH SEE ALSO
The real moon.
.SH BUGS
.B Moontool
may not always guess properly that it is running on a color Sun.  Suns with
"dual" framebuffers, like the 3/60C, can run a tool in the monochrome
plane rather than the color planes, depending on various
.B Suntools
options.  This may cause
.B moontool
to think it is running mono even though color is available.  Use the
.I -c
option to force it to try to run in color mode.
.PP
.B Moontool
uses
.B getopt(3)
to parse command line arguments, which does not interact well with some
of the standard
.B suntools(1)
arguments.  If you mix any of the
.B moontool
arguments with
.B suntools
arguments, place the
.B moontool
args first.
.SH AUTHORS
.RS 5
.nf
John Walker
http://www.fourmilab.ch/
.fi
.RE

.br
Original author of
.B moontool
(the one who did the hard part).
.PP
Ron Hitchens, Independent Hacker.  Adder-on of various "features" and
"improvements".  Culprit responsible for this stoopid man page.
.br
 ronbo@vixen.uucp
.br
 ...!uunet!cs.utah.edu!caeco!vixen!ronbo
.br
 hitchens@cs.utexas.edu


CHANGELOG
A Moon for the Unix Machine, Release 3.0

    Designed and implemented by John Walker in December 1987,
    Revised and updated in February of 1988.
    Revised and updated again in June of 1988 by Ron Hitchens.
    Revised and updated yet again in July/August of 1989 by Ron Hitchens.
    Revised and updated yet yet again in January of 2024 by Oliver Webb.

    This program was written by:

	John Walker
	http://www.fourmilab.ch/

    This  program is in the public domain: "Do what thou wilt shall be the
    whole of the law".  I'd appreciate  receiving  any  bug  fixes  and/or
    enhancements,  which  I'll  incorporate  in  future  versions  of  the
    program.  Please leave the original attribution information intact	so
    that credit and blame may be properly apportioned.

	History:
	--------
	June 1988	Version 2.0 posted to usenet by John Walker

	June 1988	Modified by Ron Hitchens to produce version 2.1
			modified icon generation to show surface texture
			on visible moon face.
  Eliminated "illegal" direct modification of icon image memory.
 added a menu to allow switching in and out of test mode, for
 entertainment value mostly.  reworked timer behaviour so that process
 doesn't wake up unnecessarily.  trap sigwinch signals to notice more
 easily when the tool opens and closes.  modified layout of information
 in open window display to reduce the amount of pixels modified in
 each update.
  Batched pixwin updates so that only one screen rasterop per cycle
  is done.
 changed open window to display white-on-black for a more aesthetic look,
 and to suggest the effect of looking at the moon in the nighttime sky.
 setup default tool and canvas colors to be the same as B&W monochrome,
 for those us lucky enough to have color monitors and who have the
 default monochrome colors set to something other than B&W (I like green
 on black myself) various code reformatting and pretty-printing to suit
 my own coding taste (I got a bit carried away).  code tweaking to make
 lint happy.  returned my hacked version to John.  (but he never got it)

	July 1989	Modified further for color displays.  On a color Sun,
			 four colors will be used for the canvas and the icon.
			 Rather than just show the illuminated portion of
			 the moon, a color icon will also show the darkened
			 portion in a dark blue shade.	The text on the icon
                         will also be drawn in a nice "buff" color, since there
			 was one more color left to use.
                        Add two command line args, "-c" and "-m" to explicitly
			 specify color or monochrome mode.  If neither are
                         given, moontool will try to determine if it's
			 running on a color or mono display by looking at the
                         depth of the frame's pixwin's pixrect.  This is not
			 always reliable on a Prism-type framebuffer like my
			 3/60C, so these two args will force one or the
			 other mode.
			Use getopt to parse the args.
			Change the tool menu slightly to use only one item
			 for switching in and out of test mode.
			A little more lint picking.

	July 1989:
  Modified a little bit more a few days later to use 8 colors and an
  accurate grey-scale moon face created by Joe Hitchens on an Amiga. Added
  the -b option to draw a one pixel border around the icon. I like it,
  but it may not mesh well with some backgrounds or adjacent icons. Added
  The Apollo 11 Commemorative Red Dot, to show where Neil and Buzz went
  on vacation a few years ago. Updated man page.
  
  August 1989: 
  Received version 2.3 of John Walker's original code.
	Rolled in bug fixes to astronomical algorithms:

  2.1  6/16/88   Bug fix.  Table of phases didn't update
	at the moment of the new moon.	Call on phasehunt didn't convert civil Julian
  date to astronomical Julian date.
	Reported by Dag Bruck (dag@control.lth.se).

  2.2  N/A	(superseded by my code)

  2.3  6/7/89	Bug fix.  Table of phases skipped the phases for July 1989. 
  This occurred due to sloppy maintenance of the synodic month index in the interchange
  of information between phasehunt() and meanphase().
  I simplified and corrected the handling of the month index as phasehunt() steps along and removed unneeded code from meanphase().  
  Reported by Bill Randle of Tektronix. (billr@saab.CNA.TEK.COM).

  I've taken it upon myself to call this version 2.4

	Ron Hitchens
		ronbo@vixen.uucp
		...!uunet!cs.utah.edu!caeco!vixen!ronbo
		hitchens@cs.utexas.edu

	December 1999	Minor Y2K fix.	Release 2.5.

  January 2024:
    Throwing out 30 year old openwindows stuff noone has or uses today
    Conversion into CLI tool
    Conversion to c99
*/


moontool.c
/* A Moon for the Unix Machine, Release 3.0

 Make with:
 cc -O2 moontool.c -o moontool -lm

 This is a program which displays the current phase of the Moon.

    The algorithms used in this program to calculate the positions Sun and
    Moon as seen from the Earth are given in the book "Practical Astronomy
    With Your Calculator" by Peter Duffett-Smith, Second Edition,
    Cambridge University Press, 1981. Ignore the word "Calculator" in the
    title; this is an essential reference if you're interested in
    developing software which calculates planetary positions, orbits,
    eclipses, and the like. If you're interested in pursuing such
    programming, you should also obtain:

    "Astronomical Formulae for Calculators" by Jean Meeus 1985. A must-have.

    "Planetary Programs and Tables from -4000 to +2800" by Pierre
    Bretagnon and Jean-Louis Simon, Willmann-Bell, 1986.

    This program was written by:

	John Walker
	http://www.fourmilab.ch/

    This  program is in the public domain: "Do what thou wilt shall be the
    whole of the law".  I'd appreciate  receiving  any  bug  fixes  and/or
    enhancements,  which  I'll  incorporate  in  future  versions  of  the
    program.  Please leave the original attribution information intact	so
    that credit and blame may be properly apportioned.
*/

#include <stdio.h>
#include <math.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/*  Astronomical constants  */
#define epoch	      2444238.5	     /* 1980 January 0.0 */
/*  Constants defining the Sun's apparent orbit  */
#define elonge	    278.833540	   /* Ecliptic longitude of the Sun at epoch 1980.0 */
#define elongp	    282.596403	   /* Ecliptic longitude of the Sun at perigee */
#define eccent      0.016718       /* Eccentricity of Earth's orbit */
#define sunsmax     1.495985e8     /* Semi-major axis of Earth's orbit, km */
#define sunangsiz   0.533128       /* Sun's angular size, degrees, at semi-major axis distance */
/*  Elements of the Moon's orbit, epoch 1980.0  */
#define mmlong      64.975464      /* Moon's mean lonigitude at the epoch */
#define mmlongp     349.383063	   /* Mean longitude of the perigee at the epoch */
#define mlnode	    151.950429	   /* Mean longitude of the node at the epoch */
#define minc        5.145396       /* Inclination of the Moon's orbit */
#define mecc        0.054900       /* Eccentricity of the Moon's orbit */
#define mangsiz     0.5181         /* Moon's angular size at distance a from Earth */
#define msmax       384401.0       /* Semi-major axis of Moon's orbit in km */
#define mparallax   0.9507	       /* Parallax at distance a from Earth */
#define synmonth    29.53058868    /* Synodic month (new Moon to new Moon) */
#define halfmonth   14.76529434    /* Half Synodic month (new Moon to full Moon) */
#define lunatbase   2423436.0      /* Base date for E. W. Brown's numbered series of lunations (1923 January 16) */
/*  Properties of the Earth  */
#define earthrad    6378.16	       /* Radius of Earth in kilometres */

#define PI 3.14159265358979323846  /* Assume not near black hole nor in Tennessee */

/*  Handy mathematical functions  */
#define sgn(x)      (((x) < 0) ? -1 : ((x) > 0 ? 1 : 0))	/* Extract sign */
#define fixangle(a) ((a) - 360.0 * (floor((a) / 360.0)))  /* Fix angle	  */
#define torad(d)    ((d) * (M_PI / 180.0))			          /* Deg->Rad	    */
#define todeg(d)    ((d) * (180.0 / M_PI))			          /* Rad->Deg	    */
#define dsin(x)     (sin(torad((x))))			                /* Sin from deg */
#define dcos(x)     (cos(torad((x))))			                /* Cos from deg */

#define resp *restrict
#define dpr double *restrict

char *phasenames[]  = { "New", "Waxing Crescent", "First Quarter", "Waxing Gibbous", "Full", "Waning Gibbous", "Last Quarter", "Waning Crescent" };
char *emojis[]      = {"🌑", "🌒", "🌓", "🌔", "🌕", "🌘", "🌗" ,"🌖"};
char *emojis_south[]= {"🌑", "🌘", "🌗", "🌖", "🌕", "🌒", "🌓" ,"🌔"};

/*  Forward functions  */
extern double phase();
void	        mprintf();
static long   jdate();
static int    phaseindex();

/* JTIME --    Convert internal GMT date and time to astronomical Julian time (i.e. Julian date plus day fraction). double jtime (struct tm *t) */
#define jtime(t) ((jdate (t) - 0.5) + (t->tm_sec + 60 * (t->tm_min + 60 * t->tm_hour)) / 86400.0)

#define HELPTXT "moontool [-h] [-t TIME] [-f FORMAT]"

int main (int argc, char **argv)
{
  time_t now = time(0);
  char *fmtstr = "%p: %e (%P%%)";

  //Option parsing
  for (int i = 0; (i = getopt (argc, argv, "ht:f:")) != -1; ) switch (i) {
    case 'h': puts(HELPTXT); exit(1);
    case 't': now = atoi(optarg); break;
    case 'f': fmtstr = strdup(optarg); break;
    default: puts("Error: Unknown Option\n"HELPTXT); exit(1);
    }

  struct tm *gm = gmtime(&now);
  double cdtd = jtime(gm), ilfrac, moonage, dist, angdia, sundist, sunangdia;
  phase (cdtd, &ilfrac, &moonage, &dist, &angdia, &sundist, &sunangdia);

  //int indx = phaseindex(ilfrac,mage);
  //printf("%s: %s (%2.1f%%)\n", phasenames[indx], emojis[indx], ilfrac*100);
  mprintf(fmtstr, ilfrac, moonage, gm);
}

// PHASEINDEX --  Return phase based on Illuminated Fraction of the moon and age of the moon
static int phaseindex (double ilumfrac, double mage)
{
  int indx;
  if (ilumfrac < 0.04) return 0;
  else if (ilumfrac > 0.96) return 4;
  else if (ilumfrac > 0.46 && ilumfrac < 0.54) return (mage > halfmonth) ? 6 : 2;
  else if (ilumfrac > 0.54) return (mage > halfmonth) ? 5 : 3;
  else return (mage > halfmonth) ? 7 : 1;
  return indx;
}

void mprintf(char *fmt, double ilumfrac, double mage, struct tm *time)
{
  int indx = phaseindex(ilumfrac, mage);
  for (int i = 0; i < strlen(fmt); i++) {
    if (fmt[i] != '%') putchar(fmt[i]);
    else switch (fmt[++i]) {
      case '\0': dprintf(2,"Error: Bad output formatting\n"); exit(1);
      case '%': putchar('%'); break;
      case 'a': printf("%2.1f", mage); break;
      case 'J': printf("%.1f", jtime(time)); break;
      case 'e': printf("%s", emojis[indx]); break;
      case 'n': putchar('\n'); break;
      case 'p': printf("%s", phasenames[indx]); break;
      case 'P': printf("%2.1f", ilumfrac*100); break;
      case 's': printf("%s", emojis_south[indx]); break;
      case 't': putchar('\t'); break;
      default: dprintf(2, "Unknown flag"); break;
    }
  }
  putchar('\n');
}

// JDATE  --  Convert internal GMT date and time to Julian day and fraction.
static long jdate (struct tm *t)
{
	long c, m, y;

	y = t->tm_year + 1900;
	m = t->tm_mon + 1;
	if (m > 2) m -= 3;
	else { m += 9; y--; }
	c = y / 100L;		   /* Compute century */
	y -= 100L * c;
	return (t->tm_mday + (c * 146097L) / 4 + (y * 1461L) / 4 + (m * 153L + 2) / 5 + 1721119L);
}


// JHMS  --  Convert Julian time to hour, minutes, and seconds.
static void jhms(double j, int resp h, int resp m, int resp s)
{
	long ij;

	j += 0.5;				/* Astronomical to civil */
	ij = (j - floor(j)) * 86400.0;
	*h = ij / 3600L;
	*m = (ij / 60L) % 60L;
	*s = ij % 60L;
}
